service/controller/v3/resource/instance/create.go

```diff
@@ -2,123 +2,488 @@ package instance
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			err = r.reimageInstance(ctx, customObject, reimagedMasterInstance, key.MasterVMSSName, key.MasterInstanceName)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "processed master VMSSs")
+		}
+	}
+
+	var allWorkerInstances []compute.VirtualMachineScaleSetVM
+	var reimagedWorkerInstance *compute.VirtualMachineScaleSetVM
+	var updatedWorkerInstance *compute.VirtualMachineScaleSetVM
+	{
+		allWorkerInstances, err = r.allInstances(ctx, customObject, key.WorkerVMSSName)
+		if IsScaleSetNotFound(err) {
+			// fall through
+		} else if err != nil {
+			return microerror.Mask(err)
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "processing worker VMSSs")
+
+			updatedWorkerInstance, reimagedWorkerInstance, err = r.nextInstance(ctx, customObject, allWorkerInstances, key.WorkerInstanceName, parameters[workerBlobKey])
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			err = r.updateInstance(ctx, customObject, updatedWorkerInstance, key.WorkerVMSSName, key.WorkerInstanceName)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			if reimagedMasterInstance == nil {
+				// In case the master instance is being updated we want to prevent any
+				// other updates on the workers. This is because the update process
+				// involves the draining of the updated node and if the master is being
+				// updated at the same time the guest cluster's Kubernetes API is not
+				// available in order to drain nodes.
+				err = r.reimageInstance(ctx, customObject, reimagedWorkerInstance, key.WorkerVMSSName, key.WorkerInstanceName)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+			} else if reimagedWorkerInstance != nil {
+				r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("not ensuring instance '%s' to be reimaged due to master processing", key.WorkerInstanceName(customObject, *reimagedWorkerInstance.InstanceID)))
+			}
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "processed worker VMSSs")
+		}
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring deployment")
+
+		masterBlobValue, err := updateVersionParameterValue(allMasterInstances, reimagedMasterInstance, key.VersionBundleVersion(customObject), parameters[masterBlobKey])
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		fmt.Printf("4: %#v\n", masterBlobValue)
+		workerBlobValue, err := updateVersionParameterValue(allWorkerInstances, reimagedWorkerInstance, key.VersionBundleVersion(customObject), parameters[workerBlobKey])
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		fmt.Printf("5: %#v\n", workerBlobValue)
+		params := map[string]interface{}{
+			masterBlobKey: masterBlobValue,
+			workerBlobKey: workerBlobValue,
+		}
+		computedDeployment, err := r.newDeployment(ctx, customObject, params)
```

Comment:
So far I tried to track the node IDs and their version bundle versions in the parameters of the new deployment introduced here. This works for new clusters created with `v3` but does not work with clusters coming from `v2` and being upgraded to `v3`. In this case the logging above causes output like this. 

```
3
4: "{\"0\":\"1.1.0\"}"
3
5: "{\"0\":\"1.1.0\",\"1\":\"1.1.0\",\"2\":\"1.1.0\"}"
```

This is wrong because we come from `v2` so it should be `1.0.0` or at least an empty string but I cannot figure how to do that. We do not have any information if there is a CR being reconciled that got its version updated. We cannot differentiate between the situation of a cluster being created with `v3` or a cluster being updated to `v3` so there is no indicator for the version blob management to get the node IDs and their version bundle versions right. With the current implementation we miss to do the actual update because the version blob now says all is up to date. One solution would be to use a different medium to track the version blob so we do not put it into the new VMSS deployment but somewhere else. Not sure what this somewhere should be then. @kopiczko WDYT?

Reply:
For the future I am thinking that it would be good to bind the version bundle version to the k8s nodes. E.g. via labels. Somebody mentioned this recently. I guess it was @teemow. When the guest cluster itself would be able to state its definite version on a node level then we would have a common ground on all providers and would not need to hack with k8s labels on pods like on KVM and hack CF stack outputs on AWS and hack whatever weird thing on Azure.

Reply:
I didn't read the code yet, but I like the idea of having that in CR status a lot. That would make the update process visible in the CR. Also has some potential to be generic.

Also maybe we can sue NodeConfig CRD for that?

Reply:
The weirdness with the status is that the operator would write and consume its own status. I am not sold that this is how it should be. I am currently favouring getting the version bound to the actual guest cluster nodes. It feels like this is how it should be. This would be the closest representation of the current state that we can get to. We have the common ground there on the interface which is the k8s guest API. So KVM, Azure and AWS would work the same way. Using the status here does not say fully "YES" to me. Should we do both? Why favouring one or the other way?

Reply:
I don't see how different storage medium can help here. And probably I don't understand the root problem. When you call `updateVersionParameterValue`, there is no `VersionParamter` then you go to this `if` https://github.com/giantswarm/azure-operator/pull/248/files#diff-ff0780da2c476a174496018b7d37a6f7R418 and the version bundle version you provide is set. But you already know that.

Would adding some code to v2 which sets the versionBlob only, help here? I think it would.

Reply:
> I don't see how different storage medium can help here.

Right now the storage medium is just something in which we write the key-value pair of node ID and version. This information is only available after some reconciliations because the templates are separated and the resources are non-blocking and the VMSS deployment has to be created in order to create nodes and so on. So we only get to know about the right key-value pairs at some point and we need indicators that tell us what kind of truth we have to record. That is why I think even the CR status would not work because it is just another storage medium in that sense and does not provide another indicator. When binding versions to the nodes themselves then the indicator is the node itself. There is nothing that can mislead the mapping between the node ID and its version then. We can just read it all the time and when the node updates its version updates and then we can read the updated information. This would also mean we are way more resilient when reconciliation breaks at some point due to whatever reason as opposed to implicitly guess based on certain information we have available during reconciliation. This is all just hacky. 

> When you call `updateVersionParameterValue`, there is no `VersionParamter` then you go to this `if` https://github.com/giantswarm/azure-operator/pull/248/files#diff-ff0780da2c476a174496018b7d37a6f7R418 and the version bundle version you provide is set. But you already know that.

I am not sure what `VersionParamter` is. I guess you mean the parameter value. The thing is that the implemented mechanism only works across multiple reconciliation loops. When we set the version blob initially the nodes are not updated because there is no information because we are just about to set it. When the version information are set we see that everything is up to date, even on an update that in fact did not get processed yet. And this is the issue. 

> Would adding some code to v2 which sets the versionBlob only, help here? I think it would.

Yes, if only the medium which we use right now is new in `v3`. In `v2` there is no separate VMSS deployment. We would need to use the specific worker and master deployments that are already there. That would work but would be a hack anyway. I would actually like to push the guest cluster node idea and have the versions bound to them in the future. I can now try to go with the dedicated deployments to just get it done and backport `v2` to make it work, but I want the node ID idea to be discussed further because right now I feel like this is the right way to go.

Reply:
> When binding versions to the nodes themselves

What do you mean by "binding versions to the nodes"?

Reply:
See https://github.com/giantswarm/giantswarm/issues/3532. We can inject the version bundle version with the user data and give it to k8s as labels or something like that.

Reply:
Yes. That would make total sense.